kv: TestTxnCoordSenderRetries failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/3cdd221e01048154f705e1efb130afb42338b462

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stressrace instead of stress and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stress TESTS=TestTxnCoordSenderRetries PKG=github.com/cockroachdb/cockroach/pkg/kv TESTTIMEOUT=5m STRESSFLAGS='-stderr=false -maxtime 20m -timeout 10m'
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=862625&tab=buildLog

```
=== RUN   TestTxnCoordSenderRetries
W180827 05:36:47.348869 32209 server/status/runtime.go:294  [n?] Could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
I180827 05:36:47.411420 32209 server/server.go:830  [n?] monitoring forward clock jumps based on server.clock.forward_jump_check_enabled
I180827 05:36:47.413863 32209 base/addr_validation.go:260  [n?] server certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180827 05:36:47.413955 32209 base/addr_validation.go:300  [n?] web UI certificate addresses: IP=127.0.0.1,::1; DNS=localhost,*.local; CN=node
I180827 05:36:47.418424 32209 server/config.go:496  [n?] 1 storage engine initialized
I180827 05:36:47.418493 32209 server/config.go:499  [n?] RocksDB cache size: 128 MiB
I180827 05:36:47.418530 32209 server/config.go:499  [n?] store 0: in-memory, size 0 B
I180827 05:36:47.479249 32209 server/node.go:373  [n?] **** cluster fb538ab1-b9e5-42ff-a996-6294d1678d5c has been created
I180827 05:36:47.479440 32209 server/server.go:1401  [n?] **** add additional nodes by specifying --join=127.0.0.1:42955
I180827 05:36:47.480367 32209 gossip/gossip.go:382  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:42955" > attrs:<> locality:<> ServerVersion:<major_val:2 minor_val:0 patch:0 unstable:12 > build_tag:"v2.1.0-alpha.20180702-2010-g3cdd221" started_at:1535348207480039898 
I180827 05:36:47.497865 32209 storage/store.go:1541  [n1,s1] [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I180827 05:36:47.498818 32209 server/node.go:476  [n1] initialized store [n1,s1]: disk (capacity=512 MiB, available=512 MiB, used=0 B, logicalBytes=6.9 KiB), ranges=1, leases=0, queries=0.00, writes=0.00, bytesPerReplica={p10=7103.00 p25=7103.00 p50=7103.00 p75=7103.00 p90=7103.00 pMax=7103.00}, writesPerReplica={p10=0.00 p25=0.00 p50=0.00 p75=0.00 p90=0.00 pMax=0.00}
I180827 05:36:47.499071 32209 storage/stores.go:242  [n1] read 0 node addresses from persistent storage
I180827 05:36:47.499416 32209 server/node.go:697  [n1] connecting to gossip network to verify cluster ID...
I180827 05:36:47.500540 32209 server/node.go:722  [n1] node connected via gossip and verified as part of cluster "fb538ab1-b9e5-42ff-a996-6294d1678d5c"
I180827 05:36:47.502287 32209 server/node.go:546  [n1] node=1: started with [<no-attributes>=<in-mem>] engine(s) and attributes []
I180827 05:36:47.503413 32209 server/status/recorder.go:652  [n1] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180827 05:36:47.503486 32209 server/server.go:1807  [n1] Could not start heap profiler worker due to: directory to store profiles could not be determined
I180827 05:36:47.511365 32492 server/status/recorder.go:652  [n1,summaries] available memory from cgroups (8.0 EiB) exceeds system memory 16 GiB, using system memory
I180827 05:36:47.513577 32209 server/server.go:1538  [n1] starting https server at 127.0.0.1:38645 (use: 127.0.0.1:38645)
I180827 05:36:47.513683 32209 server/server.go:1540  [n1] starting grpc/postgres server at 127.0.0.1:42955
I180827 05:36:47.513739 32209 server/server.go:1541  [n1] advertising CockroachDB node at 127.0.0.1:42955
W180827 05:36:47.513984 32209 jobs/registry.go:309  [n1] unable to get node liveness: node not in the liveness table
I180827 05:36:47.521735 32261 storage/replica_command.go:298  [split,n1,s1,r1/1:/M{in-ax}] initiating a split of this range at key /System/"" [r2]
I180827 05:36:47.658085 32508 storage/replica_command.go:298  [split,n1,s1,r2/1:/{System/-Max}] initiating a split of this range at key /System/NodeLiveness [r3]
W180827 05:36:47.686564 32506 storage/intent_resolver.go:668  [n1,s1] failed to push during intent resolution: failed to push "unnamed" id=22966169 key=/Table/SystemConfigSpan/Start rw=true pri=0.00367807 iso=SERIALIZABLE stat=PENDING epo=0 ts=1535348207.537424647,0 orig=1535348207.537424647,0 max=1535348207.537424647,0 wto=false rop=false seq=6
I180827 05:36:47.762201 32530 sql/event_log.go:126  [n1,intExec=optInToDiagnosticsStatReporting] Event: "set_cluster_setting", target: 0, info: {SettingName:diagnostics.reporting.enabled Value:true User:root}
I180827 05:36:47.894795 32511 storage/replica_command.go:298  [split,n1,s1,r3/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/NodeLivenessMax [r4]
I180827 05:36:48.057573 32218 storage/replica_command.go:298  [split,n1,s1,r4/1:/{System/NodeL…-Max}] initiating a split of this range at key /System/tsd [r5]
I180827 05:36:48.100268 32187 sql/event_log.go:126  [n1,intExec=set-setting] Event: "set_cluster_setting", target: 0, info: {SettingName:version Value:2.0-12 User:root}
I180827 05:36:48.166195 32192 storage/replica_command.go:298  [split,n1,s1,r5/1:/{System/tsd-Max}] initiating a split of this range at key /System/"tse" [r6]
I180827 05:36:48.252326 32553 storage/replica_command.go:298  [split,n1,s1,r6/1:/{System/tse-Max}] initiating a split of this range at key /Table/SystemConfigSpan/Start [r7]
I180827 05:36:48.291914 32190 sql/event_log.go:126  [n1,intExec=disableNetTrace] Event: "set_cluster_setting", target: 0, info: {SettingName:trace.debug.enable Value:false User:root}
I180827 05:36:48.348868 32583 storage/replica_command.go:298  [split,n1,s1,r7/1:/{Table/System…-Max}] initiating a split of this range at key /Table/11 [r8]
I180827 05:36:48.438229 32537 storage/replica_command.go:298  [split,n1,s1,r8/1:/{Table/11-Max}] initiating a split of this range at key /Table/12 [r9]
I180827 05:36:48.550083 32628 storage/replica_command.go:298  [split,n1,s1,r9/1:/{Table/12-Max}] initiating a split of this range at key /Table/13 [r10]
I180827 05:36:48.573713 32591 sql/event_log.go:126  [n1,intExec=initializeClusterSecret] Event: "set_cluster_setting", target: 0, info: {SettingName:cluster.secret Value:b88a8456-53e7-4e36-b870-607b3b6f0805 User:root}
I180827 05:36:48.649505 32573 storage/replica_command.go:298  [split,n1,s1,r10/1:/{Table/13-Max}] initiating a split of this range at key /Table/14 [r11]
I180827 05:36:48.699773 32612 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 50, info: {DatabaseName:defaultdb Statement:CREATE DATABASE IF NOT EXISTS defaultdb User:root}
I180827 05:36:48.706607 32221 rpc/nodedialer/nodedialer.go:92  [consistencyChecker,n1,s1,r1/1:/{Min-System/}] connection to n1 established
I180827 05:36:48.731565 32617 storage/replica_command.go:298  [split,n1,s1,r11/1:/{Table/14-Max}] initiating a split of this range at key /Table/15 [r12]
I180827 05:36:48.768164 32659 sql/event_log.go:126  [n1,intExec=create-default-db] Event: "create_database", target: 51, info: {DatabaseName:postgres Statement:CREATE DATABASE IF NOT EXISTS postgres User:root}
I180827 05:36:48.823570 32209 server/server.go:1594  [n1] done ensuring all necessary migrations have run
I180827 05:36:48.823713 32209 server/server.go:1597  [n1] serving sql connections
I180827 05:36:48.825338 32646 storage/replica_command.go:298  [split,n1,s1,r12/1:/{Table/15-Max}] initiating a split of this range at key /Table/16 [r13]
I180827 05:36:48.879214 32667 sql/event_log.go:126  [n1] Event: "node_join", target: 1, info: {Descriptor:{NodeID:1 Address:{NetworkField:tcp AddressField:127.0.0.1:42955} Attrs: Locality: ServerVersion:2.0-12 BuildTag:v2.1.0-alpha.20180702-2010-g3cdd221 StartedAt:1535348207480039898 LocalityAddress:[]} ClusterID:fb538ab1-b9e5-42ff-a996-6294d1678d5c StartedAt:1535348207480039898 LastUp:1535348207480039898}
I180827 05:36:48.885980 32665 server/server_update.go:67  [n1] no need to upgrade, cluster already at the newest version
I180827 05:36:48.905004 32702 storage/replica_command.go:298  [split,n1,s1,r13/1:/{Table/16-Max}] initiating a split of this range at key /Table/17 [r14]
I180827 05:36:48.981439 32724 storage/replica_command.go:298  [split,n1,s1,r14/1:/{Table/17-Max}] initiating a split of this range at key /Table/18 [r15]
I180827 05:36:49.046823 32730 storage/replica_command.go:298  [split,n1,s1,r15/1:/{Table/18-Max}] initiating a split of this range at key /Table/19 [r16]
I180827 05:36:49.086411 32716 storage/replica_command.go:298  [split,n1,s1,r16/1:/{Table/19-Max}] initiating a split of this range at key /Table/20 [r17]
I180827 05:36:49.137761 32721 storage/replica_command.go:298  [split,n1,s1,r17/1:/{Table/20-Max}] initiating a split of this range at key /Table/21 [r18]
I180827 05:36:49.179808 32760 storage/replica_command.go:298  [split,n1,s1,r18/1:/{Table/21-Max}] initiating a split of this range at key /Table/22 [r19]
I180827 05:36:49.244308 32794 storage/replica_command.go:298  [split,n1,s1,r19/1:/{Table/22-Max}] initiating a split of this range at key /Table/23 [r20]
I180827 05:36:49.336908 32751 storage/replica_command.go:298  [split,n1,s1,r20/1:/{Table/23-Max}] initiating a split of this range at key /Table/50 [r21]
I180827 05:36:49.397356 32777 storage/replica_command.go:298  [split,n1,s1,r21/1:/{Table/50-Max}] initiating a split of this range at key /Table/51 [r22]
I180827 05:36:49.452226 32209 storage/replica_command.go:298  [n1,s1,r6/1:/{System/tse-Table/System…}] initiating a split of this range at key "b" [r23]
--- FAIL: TestTxnCoordSenderRetries (2.94s)
=== RUN   TestTxnCoordSenderRetries/multi_range_DelRange_with_uncertainty_interval_error
I180827 05:36:50.172395 32860 storage/replica_command.go:73  [n1,s1,r6/1:/{System/tse-Table/System…}] test injecting error: ReadWithinUncertaintyIntervalError: read at time 1535348210.169841962,0 encountered previous write with future timestamp 1535348210.172120136,0 within uncertainty interval `t <= 1535348210.169841962,0`; observed timestamps: [{1 1535348210.169841962,0}]
I180827 05:36:50.179736 32209 util/stop/stopper.go:537  quiescing; tasks left:
1      [async] transport racer
I180827 05:36:50.179946 32502 kv/transport_race.go:91  transport race promotion: ran 20 iterations on up to 1036 requests
    --- FAIL: TestTxnCoordSenderRetries/multi_range_DelRange_with_uncertainty_interval_error (0.01s)
    	dist_sender_server_test.go:2623: expected but did not experience client retry
=== RUN   TestTxnCoordSenderRetries/multi-range_batch_with_write_too_old_and_failed_cput
    --- FAIL: TestTxnCoordSenderRetries/multi-range_batch_with_write_too_old_and_failed_cput (0.02s)
    	dist_sender_server_test.go:2607: unexpected value: raw_bytes:"\355lE\250\003value" timestamp:<wall_time:1535348210039830271 > 
=== RUN   TestTxnCoordSenderRetries/multi_range_batch_with_uncertainty_interval_error_and_mixed_success
    --- FAIL: TestTxnCoordSenderRetries/multi_range_batch_with_uncertainty_interval_error_and_mixed_success (0.00s)
    	dist_sender_server_test.go:2623: expected but did not experience client retry
=== RUN   TestTxnCoordSenderRetries/multi-range_batch_with_write_too_old_and_successful_cput
    --- FAIL: TestTxnCoordSenderRetries/multi-range_batch_with_write_too_old_and_successful_cput (0.01s)
    	dist_sender_server_test.go:2623: expected but did not experience client retry
=== RUN   TestTxnCoordSenderRetries/multi-range_batch_with_forwarded_timestamp_and_cput_and_delRange
    --- FAIL: TestTxnCoordSenderRetries/multi-range_batch_with_forwarded_timestamp_and_cput_and_delRange (0.01s)
    	dist_sender_server_test.go:2620: expected [at least] one txn coord sender auto retry; got 0

```

--- a/pkg/kv/dist_sender_server_test.go
+++ b/pkg/kv/dist_sender_server_test.go
@@ -1852,6 +1852,9 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			}
 			return nil
 		}
+	// Don't clobber the test's splits.
+	storeKnobs.DisableMergeQueue = true
+
 	s, _, _ := serverutils.StartServer(t,
 		base.TestServerArgs{Knobs: base.TestingKnobs{Store: &storeKnobs}})
 	ctx := context.Background()

--- a/pkg/kv/dist_sender_server_test.go
+++ b/pkg/kv/dist_sender_server_test.go
@@ -698,7 +698,7 @@ func TestMultiRangeBoundedBatchDelRange(t *testing.T) {
 		b.Header.MaxSpanRequestKeys = int64(bound)
 		spans := [][]string{{"a", "c"}, {"c", "f"}, {"g", "h"}}
 		for _, span := range spans {
-			b.DelRange(span[0], span[1], true)
+			b.DelRange(span[0], span[1], true /* returnKeys */)
 		}
 		if err := db.Run(ctx, b); err != nil {
 			t.Fatal(err)
@@ -751,7 +751,7 @@ func TestMultiRangeBoundedBatchDelRangeBoundary(t *testing.T) {
 
 	b := &client.Batch{}
 	b.Header.MaxSpanRequestKeys = 3
-	b.DelRange("a", "c", true)
+	b.DelRange("a", "c", true /* returnKeys */)
 	if err := db.Run(ctx, b); err != nil {
 		t.Fatal(err)
 	}
@@ -764,7 +764,7 @@ func TestMultiRangeBoundedBatchDelRangeBoundary(t *testing.T) {
 
 	b = &client.Batch{}
 	b.Header.MaxSpanRequestKeys = 1
-	b.DelRange("b", "c", true)
+	b.DelRange("b", "c", true /* returnKeys */)
 	if err := db.Run(ctx, b); err != nil {
 		t.Fatal(err)
 	}
@@ -811,7 +811,7 @@ func TestMultiRangeBoundedBatchDelRangeOverlappingKeys(t *testing.T) {
 		b.Header.MaxSpanRequestKeys = int64(bound)
 		spans := [][]string{{"a", "b3"}, {"b", "d"}, {"c", "f2a"}, {"f1a", "g"}}
 		for _, span := range spans {
-			b.DelRange(span[0], span[1], true)
+			b.DelRange(span[0], span[1], true /* returnKeys */)
 		}
 		if err := db.Run(ctx, b); err != nil {
 			t.Fatal(err)
@@ -858,8 +858,8 @@ func TestMultiRangeEmptyAfterTruncate(t *testing.T) {
 	// any active requests.
 	if err := db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
 		b := txn.NewBatch()
-		b.DelRange("a", "b", false)
-		b.DelRange("e", "f", false)
+		b.DelRange("a", "b", false /* returnKeys */)
+		b.DelRange("e", "f", false /* returnKeys */)
 		return txn.CommitInBatch(ctx, b)
 	}); err != nil {
 		t.Fatalf("unexpected error on transactional DeleteRange: %s", err)
@@ -923,7 +923,7 @@ func TestMultiRangeScanReverseScanDeleteResolve(t *testing.T) {
 	// resolved via ResolveIntentRange upon completion.
 	if err := db.Txn(ctx, func(ctx context.Context, txn *client.Txn) error {
 		b := txn.NewBatch()
-		b.DelRange("a", "d", false)
+		b.DelRange("a", "d", false /* returnKeys */)
 		return txn.CommitInBatch(ctx, b)
 	}); err != nil {
 		t.Fatalf("unexpected error on transactional DeleteRange: %s", err)
@@ -2373,7 +2373,7 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			},
 			retryable: func(ctx context.Context, txn *client.Txn) error {
 				b := txn.NewBatch()
-				b.DelRange("a", "b", false)
+				b.DelRange("a", "b", false /* returnKeys */)
 				b.CPut("c", "cput", "value")
 				return txn.CommitInBatch(ctx, b) // both puts will succeed, et will retry
 			},

--- a/pkg/kv/dist_sender_server_test.go
+++ b/pkg/kv/dist_sender_server_test.go
@@ -1658,6 +1658,9 @@ func TestPropagateTxnOnError(t *testing.T) {
 			}
 			return nil
 		}
+	// Don't clobber the test's splits.
+	storeKnobs.DisableMergeQueue = true
+
 	s, _, _ := serverutils.StartServer(t,
 		base.TestServerArgs{Knobs: base.TestingKnobs{Store: &storeKnobs}})
 	ctx := context.TODO()

--- a/pkg/kv/dist_sender_server_test.go
+++ b/pkg/kv/dist_sender_server_test.go
@@ -2490,7 +2490,7 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			clientRetry: true, // note this txn is read-only but still restarts
 		},
 		{
-			name: "multi range batch with uncertainty interval error",
+			name: "multi-range batch with uncertainty interval error",
 			beforeTxnStart: func(ctx context.Context, db *client.DB) error {
 				return db.Put(ctx, "c", "value")
 			},
@@ -2506,7 +2506,7 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			txnCoordRetry: true, // will succeed because no mixed success
 		},
 		{
-			name: "multi range batch with uncertainty interval error and get conflict",
+			name: "multi-range batch with uncertainty interval error and get conflict",
 			beforeTxnStart: func(ctx context.Context, db *client.DB) error {
 				return db.Put(ctx, "a", "init")
 			},
@@ -2528,7 +2528,7 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			clientRetry: true, // will fail because of conflict on refresh span for the Get
 		},
 		{
-			name: "multi range batch with uncertainty interval error and mixed success",
+			name: "multi-range batch with uncertainty interval error and mixed success",
 			beforeTxnStart: func(ctx context.Context, db *client.DB) error {
 				return db.Put(ctx, "c", "value")
 			},
@@ -2542,7 +2542,7 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			clientRetry: true, // client-side retry required as this will be an mixed success
 		},
 		{
-			name: "multi range scan with uncertainty interval error",
+			name: "multi-range scan with uncertainty interval error",
 			retryable: func(ctx context.Context, txn *client.Txn) error {
 				_, err := txn.Scan(ctx, "a", "d", 0)
 				return err
@@ -2551,7 +2551,7 @@ func TestTxnCoordSenderRetries(t *testing.T) {
 			txnCoordRetry: true, // can restart at higher timestamp despite mixed success because read-only
 		},
 		{
-			name: "multi range DelRange with uncertainty interval error",
+			name: "multi-range DelRange with uncertainty interval error",
 			retryable: func(ctx context.Context, txn *client.Txn) error {
 				return txn.DelRange(ctx, "a", "d")
 			},

